### ISSUE TITLE
"DB Error: no such field" when sending or viewing old Mosaico mail following Civi Upgrade 

### ISSUE THREAD

**Issue by pizzatopping** [veda-consulting/uk.co.vedaconsulting.mosaico#503]
We're seeing the error: "DB Error: no such field" when trying to send email from Mosaico. This happens regardless of whether the email is a test, or going out to the intended recipients. The error appears immediately upon submitting the mailing.

We're also unable to view old mailings submitted via Mosico, with the same error appearing but as a webpage rather than a pop-up.

This started happening after we upgraded CiviCRM from 5.39.0 to 5.46.0.

"Traditional" mails are working fine (sending and viewing)

I've done a few basic things to fix the problem with no success:

- Cleaned all caches
- Tricked Civi into thinking it was running an older version of the database and upgraded again to rule out problems with the DB update
- Upgraded Civi to 5.47.0
- Updated Mosaico to 2.9, while also installing Form Core and Seatch kit. Completed additional database updates required by this.

The error that appears while sending the mailing doesn't leave anything in the error log. I do get the following when viewing old mailings though:

```
Mar 11 16:35:20  [error] 
$Fatal Error Details = array:3 [
  "message" => "DB Error: no such field"
  "code" => null
  "exception" => CiviCRM_API3_Exception {#3486
    -extraParams: array:4 [
      "error_code" => "no such field"
      "tip" => "add debug=1 to your API call to have more info about the error"
      "is_error" => 1
      "error_message" => "DB Error: no such field"
    ]
    #message: "DB Error: no such field"
    #code: 0
    #file: ".../sites/all/modules/civicrm/api/api.php"
    #line: 134
    trace: {
      /.../sites/all/modules/civicrm/api/api.php:134 {
        › if (is_array($result) && !empty($result['is_error'])) {
        ›   throw new CiviCRM_API3_Exception($result['error_message'], CRM_Utils_Array::value('error_code', $result, 'undefined'), $result);
        › }
      }
      /.../sites/all/modules/civicrm/CRM/Mailing/Page/View.php:148 { …}
      /.../sites/all/modules/civicrm/CRM/Core/Invoke.php:319 { …}
      /.../sites/all/modules/civicrm/CRM/Core/Invoke.php:69 { …}
      /.../sites/all/modules/civicrm/CRM/Core/Invoke.php:36 { …}
      /.../sites/all/modules/civicrm/drupal/civicrm.module:471 { …}
      /.../domains/secure/htdocs/includes/menu.inc:527 { …}
      /.../domains/secure/htdocs/index.php:21 { …}
    }
  }
]


Mar 11 16:35:20  [debug] $backTrace = #0 /.../domains/secure/htdocs/sites/all/modules/civicrm/CRM/Core/Error.php(433): CRM_Core_Error::backtrace("backTrace", TRUE)
#1 /.../domains/secure/htdocs/sites/all/modules/civicrm/CRM/Core/Invoke.php(39): CRM_Core_Error::handleUnhandledException(Object(CiviCRM_API3_Exception))
#2 /.../domains/secure/htdocs/sites/all/modules/civicrm/drupal/civicrm.module(471): CRM_Core_Invoke::invoke((Array:3))
#3 /.../domains/secure/htdocs/includes/menu.inc(527): civicrm_invoke("mailing", "view")
#4 /.../domains/secure/htdocs/index.php(21): menu_execute_active_handler()
#5 {main}
```

Environment:
Debian 10 (64 bit)
MariaDB 10.3
PHP 7.3
Drupal 7.88
CiviCRM 5.47.0
Mosaico 2.9

**Comment by MarcMichalsky** [comment]
I recently ran into that "DB Error: no such field" error elsewhere. After an upgrade that added two more columns to the `civicrm_relationship` table I got the same error when I tried to add new relationships. The reason was that logging was enabled and the `log_civicrm_relationship` table did not have the new columns yet but the update and insert triggers always tried to apply all changes on the `civicrm_relationship` table also on the `log_civicrm_relationship` table. You could try to disable and enable logging again.

**Comment by GreysonStalcup** [comment]
This has also been happening for me.  Editing -> Saving mailings, attempting to send... All throw a DB Unknown Error.  I have not attempted to fix this by upgrading to 2.9 due to having the same issue listed here when attempting to upgrade prior:  https://github.com/veda-consulting-company/uk.co.vedaconsulting.mosaico/issues/504

CiviCRM: 5.47.4
Mosaico: 2.8
Drupal: 9.3.9

Backtrace / Error

```
                                                                                                                                                                                                                                                                                                                      "code" => null
  "exception" => CRM_Core_Exception {#2557
    -errorData: array:1 [
      "error_code" => 0
    ]
    #cause: null
    -_trace: null
    #message: "Could not find valid value for id"
    #code: 0
    #file: "/var/www/mysite.org/vendor/civicrm/civicrm-core/CRM/Utils/Request.php"
    #line: 98
    trace: {
      /var/www/mysite.org/vendor/civicrm/civicrm-core/CRM/Utils/Request.php:98 { …}
      /var/www/mysite.org/vendor/civicrm/civicrm-core/CRM/Event/StateMachine/Registration.php:27 { …}
      /var/www/mysite.org/vendor/civicrm/civicrm-core/CRM/Event/Controller/Registration.php:27 { …}
      /var/www/mysite.org/vendor/civicrm/civicrm-core/CRM/Core/Invoke.php:314 { …}
      /var/www/mysite.org/vendor/civicrm/civicrm-core/CRM/Core/Invoke.php:69 { …}
      /var/www/mysite.org/vendor/civicrm/civicrm-core/CRM/Core/Invoke.php:36 { …}
      /var/www/mysite.org/web/modules/contrib/civicrm/src/Civicrm.php:88 {
        Drupal\civicrm\Civicrm->invoke($args)
        › ob_start();
        › $content = \CRM_Core_Invoke::invoke($args);
        › $output = ob_get_clean();
      }
      /var/www/mysite.org/web/modules/contrib/civicrm/src/Controller/CivicrmController.php:80 { …}
      Drupal\civicrm\Controller\CivicrmController->main() {}
      /var/www/mysite.org/web/core/lib/Drupal/Core/EventSubscriber/EarlyRenderingControllerWrapperSubscriber.php:123 { …}
      /var/www/mysite.org/web/core/lib/Drupal/Core/Render/Renderer.php:564 { …}
      /var/www/mysite.org/web/core/lib/Drupal/Core/EventSubscriber/EarlyRenderingControllerWrapperSubscriber.php:124 { …}
      /var/www/mysite.org/web/core/lib/Drupal/Core/EventSubscriber/EarlyRenderingControllerWrapperSubscriber.php:97 { …}
      /var/www/mysite.org/vendor/symfony/http-kernel/HttpKernel.php:158 { …}
      /var/www/mysite.org/vendor/symfony/http-kernel/HttpKernel.php:80 { …}
      /var/www/mysite.org/web/core/lib/Drupal/Core/StackMiddleware/Session.php:58 { …}
      /var/www/mysite.org/web/core/lib/Drupal/Core/StackMiddleware/KernelPreHandle.php:48 { …}
      /var/www/mysite.org/web/core/modules/page_cache/src/StackMiddleware/PageCache.php:191 { …}
      /var/www/mysite.org/web/core/modules/page_cache/src/StackMiddleware/PageCache.php:128 { …}
      /var/www/mysite.org/web/core/modules/page_cache/src/StackMiddleware/PageCache.php:82 { …}
      /var/www/mysite.org/web/core/lib/Drupal/Core/StackMiddleware/ReverseProxyMiddleware.php:48 { …}
      /var/www/mysite.org/web/core/lib/Drupal/Core/StackMiddleware/NegotiationMiddleware.php:51 { …}
      /var/www/mysite.org/vendor/stack/builder/src/Stack/StackedHttpKernel.php:23 { …}
      /var/www/mysite.org/web/core/lib/Drupal/Core/DrupalKernel.php:708 { …}
      /var/www/mysite.org/web/index.php:19 { …}
    }
  }
]


Apr 11 02:49:32  [debug] $backTrace = #0 /var/www/mysite.org/vendor/civicrm/civicrm-core/CRM/Core/Error.php(433): CRM_Core_Error::backtrace("backTrace", TRUE)
#1 /var/www/mysite.org/vendor/civicrm/civicrm-core/CRM/Core/Invoke.php(39): CRM_Core_Error::handleUnhandledException(Object(CRM_Core_Exception))
#2 /var/www/mysite.org/web/modules/contrib/civicrm/src/Civicrm.php(88): CRM_Core_Invoke::invoke((Array:3))
#3 /var/www/mysite.org/web/modules/contrib/civicrm/src/Controller/CivicrmController.php(80): Drupal\civicrm\Civicrm->invoke((Array:3))
#4 [internal function](): Drupal\civicrm\Controller\CivicrmController->main((Array:3), "")
#5 /var/www/mysite.org/web/core/lib/Drupal/Core/EventSubscriber/EarlyRenderingControllerWrapperSubscriber.php(123): call_user_func_array((Array:2), (Array:2))
#6 /var/www/mysite.org/web/core/lib/Drupal/Core/Render/Renderer.php(564): Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->Drupal\Core\EventSubscriber\{closure}()
#7 /var/www/mysite.org/web/core/lib/Drupal/Core/EventSubscriber/EarlyRenderingControllerWrapperSubscriber.php(124): Drupal\Core\Render\Renderer->executeInRenderContext(Object(Drupal\Core\Render\RenderContext), Object(Closure))
#8 /var/www/mysite.org/web/core/lib/Drupal/Core/EventSubscriber/EarlyRenderingControllerWrapperSubscriber.php(97): Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->wrapControllerExecutionInRenderContext((Array:2), (Array:2))
#9 /var/www/mysite.org/vendor/symfony/http-kernel/HttpKernel.php(158): Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->Drupal\Core\EventSubscriber\{closure}()
#10 /var/www/mysite.org/vendor/symfony/http-kernel/HttpKernel.php(80): Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object(Symfony\Component\HttpFoundation\Request), 1)
#11 /var/www/mysite.org/web/core/lib/Drupal/Core/StackMiddleware/Session.php(58): Symfony\Component\HttpKernel\HttpKernel->handle(Object(Symfony\Component\HttpFoundation\Request), 1, TRUE)
#12 /var/www/mysite.org/web/core/lib/Drupal/Core/StackMiddleware/KernelPreHandle.php(48): Drupal\Core\StackMiddleware\Session->handle(Object(Symfony\Component\HttpFoundation\Request), 1, TRUE)
#13 /var/www/mysite.org/web/core/modules/page_cache/src/StackMiddleware/PageCache.php(191): Drupal\Core\StackMiddleware\KernelPreHandle->handle(Object(Symfony\Component\HttpFoundation\Request), 1, TRUE)
#14 /var/www/mysite.org/web/core/modules/page_cache/src/StackMiddleware/PageCache.php(128): Drupal\page_cache\StackMiddleware\PageCache->fetch(Object(Symfony\Component\HttpFoundation\Request), 1, TRUE)
#15 /var/www/mysite.org/web/core/modules/page_cache/src/StackMiddleware/PageCache.php(82): Drupal\page_cache\StackMiddleware\PageCache->lookup(Object(Symfony\Component\HttpFoundation\Request), 1, TRUE)
#16 /var/www/mysite.org/web/core/lib/Drupal/Core/StackMiddleware/ReverseProxyMiddleware.php(48): Drupal\page_cache\StackMiddleware\PageCache->handle(Object(Symfony\Component\HttpFoundation\Request), 1, TRUE)
#17 /var/www/mysite.org/web/core/lib/Drupal/Core/StackMiddleware/NegotiationMiddleware.php(51): Drupal\Core\StackMiddleware\ReverseProxyMiddleware->handle(Object(Symfony\Component\HttpFoundation\Request), 1, TRUE)
#18 /var/www/mysite.org/vendor/stack/builder/src/Stack/StackedHttpKernel.php(23): Drupal\Core\StackMiddleware\NegotiationMiddleware->handle(Object(Symfony\Component\HttpFoundation\Request), 1, TRUE)
#19 /var/www/mysite.org/web/core/lib/Drupal/Core/DrupalKernel.php(708): Stack\StackedHttpKernel->handle(Object(Symfony\Component\HttpFoundation\Request), 1, TRUE)
#20 /var/www/mysite.org/web/index.php(19): Drupal\Core\DrupalKernel->handle(Object(Symfony\Component\HttpFoundation\Request))
#21 {main}
```

**Comment by colemanw** [comment]
@GreysonStalcup what you posted is not the same error as the OP. Yours is a `"Could not find valid value for id"` error which usually results from a bad redirect to a form, or clicking refresh on a form with validation errors.

**Comment by rbaugh** [comment]
Are you using any other Mosaico extensions? I just ran into something similar to this. I had a 2.9.xxxxx version of Mosaico and pulled from the master as I was getting a weird local error on an install. When I disabled Mosaico and other extensions that depended on Mosaico, I found I was getting this same error. When I had enabled Mosaico, I had done it through CV and didn't run the extension DB upgrades. Once the Mosaico DB upgrades were run, I was able to enable the other extensions that were dependent on Mosaico.
